Add division overload with rounding_mode selection

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50438 Support JIT shape analysis for div with rounding_mode argument
* #50437 Sparse support for division rounding_mode argument
* #50281 Warn about floor_divide performing incorrect rounding
* **#50280 Add division overload with rounding_mode selection**

As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

Differential Revision: [D26123271](https://our.internmc.facebook.com/intern/diff/D26123271)

peterbell10 committed:
Add division overload with rounding_mode selection

As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

mruberry commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -1129,6 +1129,12 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
   BINARY_POINTWISE(mul);
   BINARY_POINTWISE(div);
+  {
+    using Binop = Tensor (*)(const Tensor&, const Tensor&, std::string);
```
@rzou would you take a look here?

peterbell10 replied:
@mruberry I think you got the wrong user. Was that meant to be @zou3519?

mruberry replied:
It was, thanks @peterbell10. Darn autocomplete!

cc @zou3519

zou3519 replied:
this lgtm!

mruberry commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -148,21 +150,45 @@ Tensor& copysign_(Tensor& self, Scalar other) {
   auto iter = TensorIterator::binary_float_op(result, self, other);
-  div_stub(iter.device_type(), iter);
+  div_true_stub(iter.device_type(), iter);
+  if (!result.defined()) {
```
Do the internals of this function need to change?

peterbell10 replied:
Doing it this way means I can write `div(self, other)` as
```cpp
Tensor result;
div_out(self, other, result);
return result;
```

Which means I don't have to duplicate the string comparison `if-else` chain.

mruberry replied:
OK, sounds good.

mruberry commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -179,6 +205,38 @@ Tensor& div_(Tensor& self, Scalar other) {
+    return div_floor_out(self, other, result);
+  }
+
+  AT_ERROR("div expected rounding_mode to be one of 'true', 'trunc', or 'floor' "
```
`TORCH_CHECK(false, ...)` because `AT_ERROR` is deprecated.

mruberry commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -179,6 +205,38 @@ Tensor& div_(Tensor& self, Scalar other) {
+Tensor div(const Tensor& self, const Tensor& other, std::string rounding_mode) {
+  Tensor result;
+  native::div_out(self, other, std::move(rounding_mode), result);
+  TORCH_INTERNAL_ASSERT(result.defined());
```
Probably don't need this assert?

mruberry commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -239,28 +317,14 @@ Tensor& remainder_(Tensor& self, const Tensor& other) {
-  }
-
-  return result;
+  // FIXME: Not actually doing floor division
+  return div_trunc_out(self, other, result);
```
Include the issue's url: https://github.com/pytorch/pytorch/issues/43874

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -92,24 +92,76 @@ void mul_kernel(TensorIterator& iter) {
     // There's no SIMD integer division, so don't try to vectorize it.
     // TODO: if the divisor is a scalar, rewrite as multiplication by a constant.
-    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "div_cpu", [&]() {
+    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "div_trunc_cpu", [&]() {
```
Dispatch on `iter.common_dtype()` here

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -92,24 +92,76 @@ void mul_kernel(TensorIterator& iter) {
+          // Subtracts one from the results of truncation division if the
+          // divisor and dividend have different sign(bit)s and the remainder of
+          // the division is nonzero
+          const auto quot = a / b;
```
For the CPU kernel I think this can use std::div (https://en.cppreference.com/w/cpp/numeric/math/div) but it probably doesn't matter.

peterbell10 replied:
You might be surprised but `std::div` is often slower. Compilers will fuse separate division and modulo into into a single division instruction from `-O1` and above. Meanwhile, `std::div` often isn't inlined and so has function call overheads.

https://gcc.godbolt.org/z/MhYcb6

mruberry replied:
I am surprised. Thanks for letting me know.

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -92,24 +92,76 @@ void mul_kernel(TensorIterator& iter) {
+  } else if (isIntegralType(dtype, /*includeBool*/ false)) {
+    // There's no SIMD integer division, so don't try to vectorize it.
+    // TODO: if the divisor is a scalar, rewrite as multiplication by a constant.
+    AT_DISPATCH_INTEGRAL_TYPES(iter.common_dtype(), "div_floor_cpu", [&]() {
```
This is inconsistent between using `dtype` and `iter.common_dtype()`.

peterbell10 replied:
I have removed all uses of `iter.dtype()`. If instead you meant the variable `dtype`, then I would note that it's assigned from `iter.common_dtype()` above. Just a bit less to type.

mruberry replied:
I realize the value is the same, just for readability the code might want to stick to either `dtype` or `iter.common_dtype()`. No big deal either way.

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -92,24 +92,76 @@ void mul_kernel(TensorIterator& iter) {
+      });
+    });
+  } else {
+    AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, iter.common_dtype(), "div_floor_cpu", [&]() {
```
Same `dtype` vs `iter.common_dtype` here, too.

mruberry commented on aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:
```diff
@@ -44,26 +44,100 @@ struct MulFunctor<bool> {
+    // precision compared to computing the division.
+    AT_DISPATCH_FLOATING_TYPES_AND2(kHalf, kBFloat16, dtype, "div_trunc_cuda", [&]() {
+      using accscalar_t = at::acc_type<scalar_t, true>;
+      auto inv_b = accscalar_t(1.0) / iter.scalar_value<accscalar_t>(2);
```
Would you elaborate on using the accumulate type vs. the computation type to store inv_b? In C++ this will make the result of `a * inv_b` be in the accumulate type and then it will be cast to the computation type, which seems a little odd.

peterbell10 replied:
This comes from the existing code, however I would assume it's to reduce the rounding error and avoid running out of domain from rewriting the arithmetic. It only makes a difference for half and bfloat16, all other types should be unaffected.

mruberry replied:
OK, thanks for explaining.

mruberry commented on aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:
```diff
@@ -44,26 +44,100 @@ struct MulFunctor<bool> {
+    // dividend are always the same)
+    return div_trunc_kernel_cuda(iter);
+  } else if (isIntegralType(dtype, /*includeBool*/ false)) {
+    // There's no SIMD integer division, so don't try to vectorize it.
```
This comment doesn't apply on the GPU

mruberry commented on aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:
```diff
@@ -44,26 +44,100 @@ struct MulFunctor<bool> {
+          // Subtracts one from the results of truncation division if the
+          // divisor and dividend have different sign(bit)s and the remainder of
+          // the division is nonzero
+          const auto quot = a / b;
```
fyi I don't think std::div works on both CUDA and ROCm so this pattern might be necessary for the GPU kernel

mruberry commented on aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:
```diff
@@ -44,26 +44,100 @@ struct MulFunctor<bool> {
+    // scalar, compute a * reciprocal(b). Note that this may lose one bit of
+    // precision compared to computing the division.
+    AT_DISPATCH_FLOATING_TYPES_AND2(kHalf, kBFloat16, dtype, "div_floor_cuda", [&]() {
+      using accscalar_t = at::acc_type<scalar_t, true>;
```
Same questions about accumulate type vs. computation dtype here as above.

mruberry commented on aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:
```diff
@@ -44,26 +44,100 @@ struct MulFunctor<bool> {
+      auto inv_b = accscalar_t(1.0) / iter.scalar_value<accscalar_t>(2);
+      iter.remove_operand(2);
+      gpu_kernel(iter, [inv_b] GPU_LAMBDA (scalar_t a) -> scalar_t {
+        return std::floor(a * inv_b);
```
It's a little funny that some of these cases have this optimization and others don't.

peterbell10 replied:
Can you explain? All three rounding modes have this optimization.

mruberry replied:
But there are still comments in the code suggesting it be added?

peterbell10 replied:
Ah sorry for the confusion. That comment is left from a copy-paste oversight (along with the SIMD comment above it). I'll clean it up.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1391,6 +1391,26 @@
+  variants: function, method
+  dispatch:
+    CPU, CUDA: div
+    SparseCPU, SparseCUDA: div_sparse
```
I think we can remove sparse support for rounding modes from this PR.

peterbell10 replied:
Moved sparse changes higher up the stack.

mruberry commented on test/test_torch.py:
```diff
@@ -2881,6 +2881,34 @@ def test_deepcopy_scalar(self, device, dtype):
         self.assertEqual(a, deepcopy(a))
 
+    @dtypes(*torch.testing.get_all_dtypes(include_bool=False, include_complex=False))
+    def test_div_rounding_modes(self, device, dtype):
```
Move this to test_binary_ufuncs.py.

mruberry commented on test/test_torch.py:
```diff
@@ -2881,6 +2881,34 @@ def test_deepcopy_scalar(self, device, dtype):
+        if rounding_unsupported:
+            d_true = d_true.float()
+
+        self.assertEqual(d_floor, d_true.floor().to(dtype))
```
This is good but we should explicitly compare with NumPy's floor divide and include adversarial values using two arrays that pair numbers like inf, nan, 0, -pi, pi, etc.

peterbell10 replied:
I'm adding the test and it has revealed something interesting. Python floor division isn't just `floor(a / b)`, it's more like `floor((a - fmod(a, b)) / b)` which has some weird consequences:
```python
>>> n, d = 10, 0.1; n/d, (n/d).hex(), (n//d), math.fmod(n, d).hex
(100.0, '0x1.9000000000000p+6', 99.0, 0x1.9999999999972p-4)
```
Here, `10 // 0.1 != floor(10 / 0.1)` because the result of `fmod` is isn't exact. Even though the floating point division gave the exact result. This also means `inf` doesn't propagate the same, since `fmod` results in nan when `inf` is involved.

Looking at the history, I think it's done this way to maintain the relationship that `x == (x // y) * y + (x % y)`.

mruberry replied:
This is extremely interesting. So you're saying np.floor_divide != Python's floor division? Also we should compare against Python's floor division, too.

peterbell10 replied:
Oh no, numpy uses the same definition as CPython, compare:
https://github.com/numpy/numpy/blob/ce82028409c1147a6df62d8f7437e0a9262ee2b7/numpy/core/src/npymath/npy_math_internal.h.src#L714

https://github.com/python/cpython/blob/ace008c531dd685a30c1dd68f9b5ba35f20171cf/Objects/floatobject.c#L636

I'm just highlighting some non-obvious behavior I observed and will need to change the implementation to match.

mruberry replied:
Thanks for being so thorough. We should make this clear in our documentation, too.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1391,6 +1391,26 @@
     CPU, CUDA: div_out
     SparseCPU, SparseCUDA: div_out_sparse_zerodim
 
+- func: div.Tensor_mode(Tensor self, Tensor other, str rounding_mode) -> Tensor
```
Let's make rounding_mode kwarg-only

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2610,7 +2610,14 @@ def merge_dicts(*dicts):
 
+    * ``"true"`` - no rounding and integers are promoted to floating point (default)
+                   Equivalent to the python 3 division operator: ``/``.
+    * ``"trunc"`` - truncate decimals from the true result, no type promotion.
```
"truncate decimals from the true result, no type promotion" -> "rounds the results of the division towards zero."

?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2610,7 +2610,14 @@ def merge_dicts(*dicts):
     other (Tensor or Number): the divisor
+    rounding_mode (str, optional): Type of rounding applied to the result.
 
+    * ``"true"`` - no rounding and integers are promoted to floating point (default)
```
"no rounding and integers are promoted to floating point (default)" -> "default behavior. Performs no rounding and, if both attr:\`input\` and attr:\`other\` are integer types, promotes the inputs to the default scalar type."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2610,7 +2610,14 @@ def merge_dicts(*dicts):
+                   Equivalent to the python 3 division operator: ``/``.
+    * ``"trunc"`` - truncate decimals from the true result, no type promotion.
+                    Equivalent to C-style integer division.
+    * ``"floor"`` - round down to nearest integral value, no type promotion.
```
Maybe "rounds the results of the division down."?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2610,7 +2610,14 @@ def merge_dicts(*dicts):
+    * ``"trunc"`` - truncate decimals from the true result, no type promotion.
+                    Equivalent to C-style integer division.
+    * ``"floor"`` - round down to nearest integral value, no type promotion.
+                    This is equivalent to the ``//`` operator on python numbers.
```
"Equivalent to floor division in Python (the // operator)."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2610,7 +2610,14 @@ def merge_dicts(*dicts):
+    rounding_mode (str, optional): Type of rounding applied to the result.
 
+    * ``"true"`` - no rounding and integers are promoted to floating point (default)
+                   Equivalent to the python 3 division operator: ``/``.
```
Equivalent to division in Python3 and NumPy's np.divide.

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -291,16 +291,32 @@ Tensor mul_tensor_backward(Tensor grad, Tensor other, ScalarType self_st) {
-Tensor div_tensor_self_backward(Tensor grad, Tensor other, ScalarType self_st) {
+Tensor div_tensor_self_backward(Tensor grad, Tensor other, ScalarType self_st, c10::string_view rounding_mode) {
+  if (rounding_mode != "true") {
+    return at::zeros_like(grad, grad.options().dtype(self_st));
```
@albanD to sanity-check the gradients. The floor and trunc operations do return zeros, although floor_divide simply has no gradient support today.

albanD replied:
Our story about gradients for piecewise constant functions is not the best. But as of today, returning zero gradients is the thing to do I think. In particular, it matches the behavior we already have for functions like `sign` and I would argue that consistency is the most important thing here.

mruberry commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -86,7 +87,8 @@ bool isSimpleMap(Node* node) {
       "aten::sub(Tensor self, Scalar other, Scalar alpha) -> Tensor",
       "aten::mul(Tensor self, Scalar other) -> Tensor",
-      "aten::div(Tensor self, Scalar other) -> Tensor",
+      "aten::div(Tensor self, Scalar other, str rounding_mode) -> Tensor",
```
I don't think we need to change this if this PR keeps the signature without rounding_mode?

mruberry commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -934,6 +934,7 @@ class ShapePropagator {
             "aten::sub(Tensor self, Tensor other, *, Scalar alpha) -> Tensor",
             "aten::mul(Tensor self, Tensor other) -> Tensor",
             "aten::div(Tensor self, Tensor other) -> Tensor",
+            "aten::div(Tensor self, Tensor other, str rounding_mode) -> Tensor",
```
Do we need to change this to pass a test? If not, I suggest we don't edit it. It's not well tested.

mruberry commented on torch/csrc/jit/runtime/symbolic_script.cpp:
```diff
@@ -761,6 +761,31 @@ const std::vector<std::string> functions = {
                 return grad_output / other, None
             return self / other, backward
 
+        def div_2(self, other, rounding_mode: str):
```
Are these necessary?

peterbell10 replied:
If I remove it the autodiff test fails, if I remove `assert_autodiffed` from the `OpInfo` then I instead get this error:
```
E   AssertionError: False is not true : 
E   Failure in testing nodes' autodifferentiation, one or more nodes were not expected to be autodiffed, but were found in a fused/nonfused DifferentiableGraph group. Did you intend for these nodes to be autodiffed? If so, change this test to expect autodifferention. 
E   Specifically:
E     aten::div was not expected to be in a DifferentiableGraph but was.
```
Because some sample inputs aren't autodiffed while others are.

mruberry replied:
This goes back to my previous question, which is whether we should model this as multiple operators for testing?

peterbell10 replied:
So I should have 3 `OpInfos`: no `rounding_mode`, `rounding_mode="true"` and `rounding_mode={"floor", "trunc"}`?

mruberry replied:
That seems more straightforward than these other workarounds to me, but I could be convinced it's easier to create one OpInfo (or two, or four) if you think that's best.

peterbell10 replied:
I split the `OpInfo`s up by `rounding_mode` but it was still inconsistent even for the same rounding mode. So, I think what's happening is the tensor overload is implicitly autodiff-able but the scalar version has to be defined manually.

mruberry replied:
Interesting... let me know if it would be helpful to chat offline about the details here. I'd like to find a simple approach to address this.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -350,6 +350,30 @@ def sample_inputs_broadcast_to(op_info, device, dtype, requires_grad):
                  for size, shape in test_cases)
 
+def sample_inputs_div(self, device, dtype, requires_grad):
+    a = make_tensor((S, S, S), device, dtype, low=None, high=None, requires_grad=requires_grad)
```
It's unfortunate that make_tensor doesn't currently around for two separate domains to pull values from. For types that support negative values we should add a second "b" tensor to pair a with that contains negative values.

mruberry reviewed (commented):
Hey @peterbell10!

This is an exciting PR. I've made some suggestions and asked a few questions inline. The sparse and jit improvements can probably be dropped to simplify this. 

Looking forward to hearing your thoughts.

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

peterbell10 commented:
XLA build failures seem to be from XLA's codegen. It has some special handling div but must not filter by signature in any way:
https://github.com/pytorch/xla/blob/1d691099edbc083ef51d402a87d8dca9148ea645/scripts/gen.py#L120

Not sure how to fix that.

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

mruberry commented:
> XLA build failures seem to be from XLA's codegen. It has some special handling div but must not filter by signature in any way:
> https://github.com/pytorch/xla/blob/1d691099edbc083ef51d402a87d8dca9148ea645/scripts/gen.py#L120
> 
> Not sure how to fix that.

I wonder if the dispatch keys need to changed. @ailzhang, @JackCaoG, we're adding a kwarg-only argument to torch.div() and torch.divide() while keeping the existing signatures. Would you take a look at the error we're getting and the current native_functions.yaml signatures? Maybe more of them need to be `CPU, CUDA` dispatches?

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

[ghstack-poisoned]

mruberry commented:
Hey @peterbell10 and @JackCaoG. Looks like @JackCaoG has an XLA PR we can coordinate for when this is ready to merge, is that right, @JackCaoG? Thanks for always being so responsive on these issues.

@peterbell10 Is this ready for another review?

JackCaoG commented:
> Hey @peterbell10 and @JackCaoG. Looks like @JackCaoG has an XLA PR we can coordinate for when this is ready to merge, is that right, @JackCaoG? Thanks for always being so responsive on these issues.
> 
> @peterbell10 Is this ready for another review?

@mruberry yup, I will try to get approval for the xla pr, but it should fix the test failure.

peterbell10 commented:
@mruberry this should be good for another round of review.

@JackCaoG just FYI, this PR implements Python's definition of floor division which is more complicated than just `floor(a/b)` and will round differently for inputs where `fmod` is inaccurate.

JackCaoG commented:
Hi @peterbell10, could you give a bit more detail about the difference between python floor and  `floor(a/b)`? Some example would be super helpful! Thanks.

peterbell10 commented:
@JackCaoG Python `floor_division` is defined based on the expression `(a - fmod(a, b)) / b`. Plus some fix-ups which you can see in the CPython implementation:
https://github.com/python/cpython/blob/ace008c531dd685a30c1dd68f9b5ba35f20171cf/Objects/floatobject.c#L636

An example where this matters is `10. // 0.1 == 99` even though `math.floor(10 / 0.1) == 100`. This happens because `math.fmod(10, 0.1) == 0.09999999999999945` instead of 0.  Mainly, I think this happens when the result is almost - but not exactly - divisible. Maybe that small discrepancy is unavoidable though.

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -92,29 +92,106 @@ void mul_kernel(TensorIterator& iter) {
+      });
+    });
+  } else {
+    // NOTE: This round-about way of calculating floor(a / b) is needed for exact python compatibility
```
Create a Note describing how Python implements floor division. The note can appear above the operation and be referenced from here.

mruberry commented on aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:
```diff
@@ -44,26 +44,129 @@ struct MulFunctor<bool> {
+    // optimization for floating-point types: if the second operand is a CPU
+    // scalar, compute a * reciprocal(b). Note that this may lose one bit of
+    // precision compared to computing the division.
+    AT_DISPATCH_FLOATING_TYPES_AND2(kHalf, kBFloat16, dtype, "div_floor_cuda", [&]() {
```
Reference the [Floor Division in Python] note here, too

mruberry commented on aten/src/ATen/native/cuda/Loops.cuh:
```diff
@@ -91,7 +91,9 @@ template <typename func_t>
-    TORCH_INTERNAL_ASSERT(iter.device(arg).is_cuda());
+    TORCH_INTERNAL_ASSERT(
+      iter.device(arg).is_cuda(),
+      "argument ", arg, ": expected device to be cuda but found ", iter.device(arg));
```
"expected device to be cuda" -> "expected a CUDA device"

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1417,6 +1446,25 @@
+  variants: method
+
+- func: divide.out_mode(Tensor self, Tensor other, *, str rounding_mode, Tensor(a!) out) -> Tensor(a!)
+  use_c10_dispatcher: hacky_wrapper_for_legacy_signatures
```
This can probably ditch the hacky wrapper.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -268,6 +268,72 @@ def test_inplace_division(self, device):
+        if dtype.is_floating_point:
+            a = 10 * torch.randn(100, device=device, dtype=dtype)
+            # Avoid division by zero so we can test (a / b) * b == a
+            b = 0.1 + 2 * torch.rand(100, device=device, dtype=dtype)
```
Use make_tensor here and above. To avoid zeros just replace the zeros in the tensor with another value, like 1.1.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -268,6 +268,72 @@ def test_inplace_division(self, device):
+        else:
+            info = torch.iinfo(dtype)
+            low, high = info.min, info.max
+            a = torch.randint(low, high, (100,), device=device, dtype=dtype)
```
make_tensor might allow you to avoid two branches

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -268,6 +268,72 @@ def test_inplace_division(self, device):
+        rounding_unsupported = (
+            dtype == torch.half and device != 'cuda' or
+            dtype == torch.bfloat16 and device != 'cpu')
+        d_ref = d_true.float() if rounding_unsupported else d_true.float()
```
This always becomes d_true.float()?

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -268,6 +268,72 @@ def test_inplace_division(self, device):
+        self.assertEqual(d_trunc, d_ref.trunc().to(dtype))
+
+        # Compare against NumPy
+        if dtype.is_floating_point:
```
For the NumPy comparison start a new test and set the default scalar type to double in a try/finally block and test uint8, int8, long, float32, and complex64 dtype division. Enumerate the modes on large random tensors (with zero divisors removed for the integer types), and then every other element of those tensors (similar to what you're doing below). 

This should simplify the edge cases and test more values. The tensors can be of size 4096, for example.

peterbell10 replied:
>For the NumPy comparison start a new test and set the default scalar type to double in a try/finally block and test uint8, int8, long, float32, and complex64 dtype division.

Split this into 2 tests. `div_rounding_nonfinite` to test fp special values and `div_rounding_numpy` to test division of random tensors. Note that the rounding modes don't support complex. This is consistent with Python, e.g.:
```python
>>> 1j // 2
TypeError: can't take floor of complex number.
```

Although, I'm not sure what you meant by 

> Enumerate the modes on large random tensors [...]  **and then every other element of those tensors**

Not sure what you mean in the second part. Do you mean `divide(a[None, :], b[:, None], ...)` so I'm testing 4096 * 4096 combinations?

mruberry replied:
> @mruberry I'm not sure what you meant by the second part. Do you mean `divide(a[None, :], b[:, None], ...)` so I'm testing 4096 * 4096 combinations?

Sorry, that was unclear. I was thinking:

```
a = ...
a_discontiguous = a[::2]
```

Then a_discontiguous will have every other element of a. Does that make sense?

peterbell10 replied:
Oh I see, "every second element" not "all other elements".

mruberry commented:
Hey @peterbell10, had a chance to take a look again and I made a few small suggestions. This PR is incredible. Your attention to detail and the nuance in some of these changes is exceptional. 

Aside from these minute suggestions, is this PR ready, @peterbell10? 

@JackCaoG, we're hoping to merge this before the 1.8 branch cut scheduled for February 4th. Do you think that will be an issue? If it is, is there some way to work around the PyTorch/XLA problem? Possibly by just throwing a runtime error with PyTorch/XLA if a user attempts to use div with a rounding mode?

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

Differential Revision: [D26123271](https://our.internmc.facebook.com/intern/diff/D26123271)

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

Differential Revision: [D26123271](https://our.internmc.facebook.com/intern/diff/D26123271)

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

Differential Revision: [D26123271](https://our.internmc.facebook.com/intern/diff/D26123271)

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

Differential Revision: [D26123271](https://our.internmc.facebook.com/intern/diff/D26123271)

[ghstack-poisoned]

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

Differential Revision: [D26123271](https://our.internmc.facebook.com/intern/diff/D26123271)

[ghstack-poisoned]

JackCaoG commented:
Hi @mruberry, I think I can get pt/xla pr ready this week. I will ping you when that is ready.

peterbell10 commented:
@mruberry PTAL when you can. Have addressed your comments and rebased.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -268,6 +268,122 @@ def test_inplace_division(self, device):
+            self.assertEqual(actual, torch.from_numpy(expect),
+                             exact_device=False, exact_dtype=exact_dtype)
+
+        # Compare contiguous (likely vectorized) against non-contiguous (not vectorized)
```
Is this part of the test now redundant with the following test?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2719,33 +2719,51 @@ def merge_dicts(*dicts):
+        * ``"true"`` - default behavior. Performs no rounding and, if both :attr:`input` and
+          :attr:`other` are integer types, promotes the inputs to the default scalar type.
+          Equivalent to true division in Python (the ``/`` operator) and NumPy's ``np.true_divide``.
+        * ``"trunc"`` - rounds the results of the division down.
```
The description for trunc and floor are identical, "rounds the results of the division down". 

For trunc I think we can say rounds towards zero?

https://github.com/pytorch/pytorch/pull/50280#discussion_r555070519

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -641,8 +641,10 @@ def freeze_rng_state():
     torch.set_default_dtype(dtype)
-    yield
-    torch.set_default_dtype(saved_dtype)
+    try:
```
Thank you for fixing this.

mruberry reviewed (approved):
One doc nit, otherwise looks awesome.

@JackCaoG let us know when this is safe to merge.

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

Differential Revision: [D26123271](https://our.internmc.facebook.com/intern/diff/D26123271)

[ghstack-poisoned]

mruberry commented:
One of the test failures is real: test_div_rounding_numpy_cuda_bfloat16.

We can skip the test for simplicity to unblock this landing. @JackCaoG's PR is ready to go so I'd like to land this during PST business hours on Tuesday, February 2nd.

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

Differential Revision: [D26123271](https://our.internmc.facebook.com/intern/diff/D26123271)

[ghstack-poisoned]

peterbell10 commented:
In hindsight, the BFloat16 comparison with random test data is unlikely to work perfectly. It's sensitive to exact rounding and since NumPy is rounding to float32 precision, it will occasionally get different answers.

mruberry commented:
> In hindsight, the BFloat16 comparison with random test data is unlikely to work perfectly. It's sensitive to exact rounding and since NumPy is rounding to float32 precision, it will occasionally get different answers.

We can disable the test for now. In the future we could use a fixture to ensure there are no rounding issues.

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -92,29 +92,116 @@ void mul_kernel(TensorIterator& iter) {
+            }
+            return floordiv;
+          },
+          [](Vec256<scalar_t> a, Vec256<scalar_t> b) -> Vec256<scalar_t>{
```
This is triggering some internal build issues. Adding a vectorized function can be a little tricky because we often have to stub them out on some platforms, like Android.

Since we're so close to the branch cut, I propose removing the copysign implementation and this vectorized implementation. We can file an issue and add them back in a later PR where we can take our time and focus on that issue.

peterbell10 replied:
@mruberry this should be good now. Removed all Vec256 changes and unvectorized `floor_divide`.

peterbell10 committed:
Update on "Add division overload with rounding_mode selection"


As mentioned in gh-43874, this adds a `rounding_mode={'true', 'trunc', 'floor'}`
argument so `torch.div` can be used as a replacement for `floor_divide` during
the transitional period.

I've included dedicated kernels for truncated and floor division which
aren't strictly necessary for float, but do perform significantly better (~2x) than
doing true division followed by a separate rounding kernel.

Note: I introduce new overloads for `aten::div` instead of just adding a default
`rounding_mode` because various JIT passes rely on the exact operator schema.

Differential Revision: [D26123271](https://our.internmc.facebook.com/intern/diff/D26123271)

[ghstack-poisoned]

Commit:
trial fix for mobile manifest issue

mruberry commented:
FYI there's another internal issue that I'm reviewing now. I'll keep this updated.

mruberry commented:
Update: hacking through internal infra issues still.

mruberry commented:
Update: blocking infra team confirms it's fixed it's issue. This should land today.

mruberry commented:
Landed. Some changes had to be made internally:

- div and floor_divide had to call the stub directly and not re-dispatch to the out= variant.
- the kernel name "div_cpu" couldn't change because mobile manifests list kernel names for export, and they didn't understand the new name
